### DISCUSSION TITLE
Trying to pmap a list of PyTree

### DISCUSSION THREAD

**Discussion by ThatBitBoss** [jax-ml/jax#19133] [Q&A]
[Here is my code](https://colab.research.google.com/drive/19W8lWc5QcDajag5Jkj8-PQ1Ph9N98gH6#scrollTo=rLiHnhbgvjcR&line=1&uniqifier=1)

When I go over an[ array of integers - it seems to work](https://colab.research.google.com/drive/19W8lWc5QcDajag5Jkj8-PQ1Ph9N98gH6#scrollTo=0lPueOFaqc4i&line=1&uniqifier=1)

But the list of PyTree seems to throw an error.
Ultimately I want to pmap to a [execusteGoal_Hunger](https://colab.research.google.com/drive/19W8lWc5QcDajag5Jkj8-PQ1Ph9N98gH6#scrollTo=4pJT8InpqUqe&line=3&uniqifier=1) and I am obviously missing or messing up something.

**Comment by jakevdp** [comment]
Hi - thanks for the question! `pmap` operates over arrays, not over lists. When collections like lists, tuples, or dicts are passed to JAX functions like `pmap`, `vmap`, `map`, `scan`, etc. the semantics are that these function iterate over *all arrays in the collection*.

So to do what you wish here, instead of passing a list of pytrees, you should pass a pytree of arrays. Here's a simple example of what I mean, using `vmap` for simplicity:
```python
import jax

def f(x):
  return x['A'] + x['B']

list_of_trees = [
    {'A': 1, 'B': 2},
    {'A': 5, 'B': 10}
]
jax.vmap(f)(list_of_trees)  # wrong!
# ValueError: vmap was requested to map its argument along axis 0, which implies
#             that its rank should be at least 1, but is only 0 (its shape is ())

tree_of_arrays = {'A': jnp.array([1, 5]), 'B': jnp.array([2, 10])}
result = jax.vmap(f)(tree_of_arrays)  # correct!
print(result)
# [ 3 15]
```
Your code uses a `list_of_trees` pattern, instead you should use a `tree_of_arrays` pattern if you want the function to be compatible with `pmap` or `vmap`.

Hope that helps!
